src/test/app/Batch_test.cpp

```diff
@@ -0,0 +1,2148 @@
+            BEAST_EXPECT(env.balance(alice, USD.issue()) == preAliceUSD);
+            BEAST_EXPECT(env.seq(bob) == preBobSeq);
+            BEAST_EXPECT(env.balance(bob) == preBob);
+            BEAST_EXPECT(env.balance(bob, USD.issue()) == preBobUSD);
+        }
+
+        // Invalid: Outer and Inner Sequence are the same
+        {
+            auto const preAliceSeq = env.seq(alice);
+            auto const preAlice = env.balance(alice);
+            auto const preAliceUSD = env.balance(alice, USD.issue());
+            auto const preBobSeq = env.seq(bob);
+            auto const preBob = env.balance(bob);
+            auto const preBobUSD = env.balance(bob, USD.issue());
+
+            auto const batchFee = calcBatchFee(env, 1, 2);
+            env(batch::batch(alice, preAliceSeq, batchFee, tfAllOrNothing),
+                batch::add(pay(alice, bob, XRP(10)), preAliceSeq),
+                batch::add(pay(bob, alice, XRP(5)), preBobSeq),
+                batch::sig(bob),
+                ter(tesSUCCESS));
+            TxID const batchId = env.tx()->getTransactionID();
+            std::vector<TestBatchData> testCases = {};
+            env.close();
+            validateBatch(env, batchId, testCases);
+
+            // Alice pays fee & Bob should not be affected.
+            BEAST_EXPECT(env.seq(alice) == preAliceSeq + 1);
+            BEAST_EXPECT(env.balance(alice) == preAlice - batchFee);
+            BEAST_EXPECT(env.balance(alice, USD.issue()) == preAliceUSD);
+            BEAST_EXPECT(env.seq(bob) == preBobSeq);
+            BEAST_EXPECT(env.balance(bob) == preBob);
+            BEAST_EXPECT(env.balance(bob, USD.issue()) == preBobUSD);
+        }
+    }
+
+    void
+    testBadFeeOuterBatch(FeatureBitset features)
+    {
+        testcase("bad fee outer batch");
+
+        using namespace test::jtx;
+        using namespace std::literals;
+
+        // Bad Fee Without Signer
+        {
+            test::jtx::Env env{*this, envconfig()};
+            XRPAmount const feeDrops = env.current()->fees().base;
+
+            auto const alice = Account("alice");
+            auto const bob = Account("bob");
+            auto const gw = Account("gw");
+            auto const USD = gw["USD"];
+
+            env.fund(XRP(1000), alice, bob, gw);
+            env.close();
+            env.trust(USD(1000), alice, bob);
+            env(pay(gw, alice, USD(100)));
+            env(pay(gw, bob, USD(100)));
+            env.close();
+
+            env(noop(bob), ter(tesSUCCESS));
+            env.close();
+
+            auto const preAliceSeq = env.seq(alice);
+            auto const preAlice = env.balance(alice);
+            auto const preAliceUSD = env.balance(alice, USD.issue());
+            auto const preBobSeq = env.seq(bob);
+            auto const preBob = env.balance(bob);
+            auto const preBobUSD = env.balance(bob, USD.issue());
+
+            auto const batchFee = (0 + 1) * feeDrops;
```

Comment:
The tests here would be a little better if they tested `correctFee - 1` instead of doing different math.